tasks/test_results_processor.py

```diff
@@ -0,0 +1,200 @@
+import base64
+import json
+import logging
+import random
+import zlib
+from copy import deepcopy
+from datetime import datetime
+from io import BytesIO
+from json import loads
+from typing import List
+
+from redis.exceptions import LockError
+from shared.celery_config import test_results_processor_task_name
+from shared.config import get_config
+from shared.yaml import UserYaml
+from sqlalchemy.orm import Session
+from testing_result_parsers import (
+    Testrun,
+    parse_junit_xml,
+    parse_pytest_reportlog,
+    parse_vitest_json,
+)
+
+from app import celery_app
+from database.models import Repository, Upload
+from database.models.reports import Test, TestInstance
+from services.archive import ArchiveService
+from services.yaml import read_yaml_field
+from tasks.base import BaseCodecovTask
+
+log = logging.getLogger(__name__)
+
+
+class TestResultsProcessorTask(BaseCodecovTask, name=test_results_processor_task_name):
+    __test__ = False
+
+    async def run_async(
+        self,
+        db_session,
+        *,
+        repoid,
+        commitid,
+        commit_yaml,
+        arguments_list,
+        report_code=None,
+        **kwargs,
+    ):
+        commit_yaml = UserYaml(commit_yaml)
+
+        repoid = int(repoid)
+        log.info(
+            "Received upload test result processing task",
+            extra=dict(repoid=repoid, commit=commitid),
+        )
+
+        testrun_list = []
+        upload_list = []
+
+        # process each report session's test information
+        for args in arguments_list:
+            upload_obj: Upload = (
+                db_session.query(Upload).filter_by(id_=args.get("upload_pk")).first()
+            )
+
+            try:
+                parsed_testruns: List[Testrun] = self.process_individual_arg(
+                    upload_obj, upload_obj.report.commit.repository
+                )
+            except Exception:
+                log.warning(f"Error parsing testruns")
+                return {"successful": False}
+
+            # concat existing and new test information
+            testrun_list += parsed_testruns
```

Comment:
Looking at the `upload` task it seems that you decided to go with a `group`, that is parallel in nature (meaning we might be processing multiple uploads at the same time for the same commit).

This indicates that it's ok to be saving data concurrently for the same commit. I wonder then if it's necessary to accumulate the parsing results in `testrun_list` prior to saving them. I think this list might become possibly big (we know of users that have +100K tests, for example), and this seems like it could use a fairly large amount of data.

Reply:
I could save the instances as soon as we parse them, but for tests I think it's better to avoid doing it on every iteration since we will be getting the list of all tests for that repository